### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by willy2dg**
Created sound wave view interface based to solve #1 .

### CODE REVIEW

**Comment by nkt** [comment]
Hey, thanks for PR!
Please convert your code to kotlin and add tests.

**Comment by willy2dg** [comment]
The code is in Kotlin and the repository is a test application, so that the result can be displayed, the module itself is in the sound waves folder. I will add some unit functionality tests anyway.

**Comment by willy2dg** [comment]
Hi @nkt , i improved and extended the README file. Please, review the PR when possible.

**Comment by inwady** [comment]
Hello, @willy2dg, sorry for the wait.

Resources is calculated very long time by Google Pixel 3. Look at http://bit.ly/2L9s7KN. You're able to change that, right?

Several comments:
* could you remove the all `!!` operators from your code because it's not best practise in Koltin?
* may it be better to write one universal function for decode (SoundDecoder.kt#L21 and SoundDecoder.kt#L48)?

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    <application
+            android:allowBackup="true"
+            android:icon="@mipmap/ic_launcher"
+            android:label="@string/app_name"
+            android:roundIcon="@mipmap/ic_launcher_round"
+            android:supportsRtl="true"
+            android:theme="@style/AppTheme">
+
+        <activity android:name=".MainActivity"
+                  android:configChanges="orientation|screenSize|keyboardHidden"
```
Comment (inwady) [inline app/src/main/AndroidManifest.xml]: Could you develop View without this attribute in Manifest?

**Comment by willy2dg** [comment]
Hi @inwady , thanks for review the PR. Let my clarify the points.

- I will remove all  `!!` operators
- The `decode` functions you mention are just wrappers for the "universal" one, they are required because `assets` ( sounds included in the app at build time ) are not files in the device, so i just added a function which allow an asset as input. It can be removed but doesnt speed up the function.
- The attribute you point in the Manifest is in the test app ( not in the module ), so you can create your app without it, but then the view can't be rotated on orientation changes.

About speed up, i can suggest/implement some speed up's but decoding the sound and parsing the data will take some time any way, since it is a expensive task. So provably an activity indicator could do the trick, i can implement a interface for it, but including some one should be out of the issue.

- Negative wave ( bottom part of the wave if displayNegative is true ), in the module it is calculated to display the real negative wave, while many implementations just get the positive one and mirror it. I can do the same to avoid more computation in this case, i am sure it will help.
- I will modify the `SoundGrouper` interface to handle all blocks in a call, this will reduce function calls slightly growing this function parameters.
- I will try to include the compute block in async task, it will take the same time to do the job, but next `SoundWave` instances will begin to decode early if device just allow a decoder instance.

Can we close the issue once i included all this fixes? Do you have some more concerns/questions?

**Comment by willy2dg** [comment]
@nkt , @inwady I pushed a update with mentioned changes

**Comment by inwady** [comment]
Hi! Thank you for your fixes!

I tried to use the library in my code. There're issues:
* Speed is main part of View. Maybe you be able to reduce count of samples?
* I found bug — try to move the current track when music is played. Fixed it: [crutch, need to change](https://drive.google.com/file/d/1ErfrWzRLzfgoLUDaiRwglfdSpFgkiD3M/view?usp=sharing)
* I found out that View don't save state after changing Activity configuration. We cannot approve this code due to "configChanges".

**Comment by willy2dg** [comment]
Hi @inwady , thanks for you quick review.

- I understand you want it as fast as posible, but this is a task can't be done "instantly". I will try to skip some samples, but i don't think it helps a lot, because accessing buffers by index will be slower than reading it continously.
- The app implementation was just a example to review the code, but i will be adding your fixes.
- I will implement the onSaveStateChanges, but without "confiChanges" you will need to recompute the sound wave again if config changes, since you can't store many data in this way.

**Comment by inwady** [comment]
Well! We look forward to further fixes.

**Comment by willy2dg** [comment]
Hi @inwady, I sent a new commit, I reduced the amount of decoded samples to handle, it reduces the loading time slightly and allows us to cache the samples when the state is saved. 
I removed the configChanges from the manifest, and now the view saves his state.
I also added your fixes for the example app. 
Hopefully we can close it now, wait for your feedback.

**Comment by inwady** [comment]
Hello, thanks, I checked it out.

Could you add basic listener in `SimplePlayer` which ll contain events:

- when music is started
- when music is finished

In order `SimplePlayer` could draw`R.drawable.ic_action_pause` when music is really stopped by your library.

And please reformat code (delete unused code and etc) for merge.

**Comment by willy2dg** [comment]
I sent that changes and updated README.md

**Pull request merged**